src/main/java/shef/servlets/TestDisplayRecipesServlet.java

```diff
@@ -0,0 +1,86 @@
+// limitations under the License.
+ 
+package shef.servlets;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import shef.data.Recipe;
+import shef.data.Step;
+import shef.servlets.NewRecipeServlet;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+ 
+/** Servlet responsible for displaying comments. */
+@WebServlet("/display-recipes")
+public class TestDisplayRecipesServlet extends HttpServlet {
+ 
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Recipe").addSort("timestamp", SortDirection.DESCENDING);
+ 
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+ 
+    List<Recipe> recipes = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      recipes.add(entityToRecipe(entity));
+    }
+ 
+    Gson gson = new Gson();
+ 
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(recipes));
+  }
+
+   /** Converts a Datastore entity into a Recipe. */
+  public Recipe entityToRecipe(Entity recipeEntity) {
+    String key = KeyFactory.keyToString(recipeEntity.getKey());
+    String name = (String) recipeEntity.getProperty("name");
+    String description = (String) recipeEntity.getProperty("description");
+    LinkedHashSet<String> tags = new LinkedHashSet<>((LinkedList<String>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("tags"), "tag"));
+    LinkedHashSet<String> ingredients = new LinkedHashSet<>((LinkedList<String>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("ingredients"), "ingredient"));
+    LinkedList<Step> steps = (LinkedList<Step>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("steps"), "step");
+    long timestamp = (long) recipeEntity.getProperty("timestamp");
+    return new Recipe(key, name, description, tags, ingredients, steps, timestamp);
+  }
+
+  /** Gets a list of Recipe parameters from a Datastore property. */
+  public Collection<Object> getDataAsList(Object propertiesObject, String field) {
```

Comment:
@tgomezzzz can i make this public static in Recipe..?

Reply:
for entityToRecipe, there are subtle differences, but this method is exactly the same.